### ISSUE TITLE
libs/libxercesc needs help with Linux and Mac builds

### ISSUE THREAD

**Issue by Rasbats** [Rasbats/RTZreader_pi#1]
Cannot solve the problem of includes.
@PATG : Could you help with these?

**Comment by Rasbats** [comment]
Failed builds are here:
[https://app.circleci.com/pipelines/github/Rasbats/RTZreader_pi](https://app.circleci.com/pipelines/github/Rasbats/RTZreader_pi)

**Comment by leamas** [comment]
The plugin uses the xerces parser to parse some xml. Actually, some c implementation (the original xerces is java)

Anyway: two solutions: Add xercesc to the libs, or use for example pugixml instead. The latter is no doubt the best, but will require some code  changes.

I would strongly advice against adding xerces in any form to opencpn-libs, pugixml  does a far better job.  We don't need yet another XML parser. That is not to say the plugin cannot bundle it ~itself~  a private copy.

**Comment by Rasbats** [comment]
The plugin uses pugixml for all the parsing. I need to validate against the _RTZ .xsd_. Not sure if pugixml can do that?
I am trying to use xercesc from a private _libs/libxercesc_, which works for Windows. Just cannot find the includes for Linux/Mac.
This is how @PATG made it work for a standalone app:
[https://github.com/Rasbats/opencpn.rtz.assist/blob/master/CMakeLists.txt](https://github.com/Rasbats/opencpn.rtz.assist/blob/master/CMakeLists.txt)
Cannot make this work from my private _libs/libxercesc_ addition.

**Comment by Rasbats** [comment]
[https://github.com/zeux/pugixml/issues/407](https://github.com/zeux/pugixml/issues/407)

**Comment by leamas** [comment]
Ok, I see.

> I am trying to use xercesc from a private libs/libxercesc, which works for Windows. Just cannot find the includes for Linux/Mac

Do use a precompiled library, or build it from source? 

In the latter case it's just some trivial CMake coding in the library so linking the target also supplies the headers. Furthermore, this would then not depend on platform.

It's hard, really hard to keep a precompiled library available and up do date with all supported linux and macos targets. Not impossible,  but a lot of work which in almost all cases is hard to justify.

**Comment by Rasbats** [comment]
CMake is finding the headers as well as the lib:
```
-- Found XercesC: /usr/lib/x86_64-linux-gnu/libxerces-c.so (found version "3.2.3")
-- include_directories (XercesC_INCLUDE_DIR: /usr/include)
```
But this bit is not working:
```
include_directories(_XERCES INTERFACE ${XercesC_INCLUDE_DIR})
  add_library(ocpn::xercesc ALIAS _XERCES)
```

**Comment by leamas** [comment]
> Found XercesC: /usr/lib/x86_64-linux-gnu/libxerces-c.so (found version "3.2.3")

No, big no. This means that the plugin would depend on  an external library, which means that the end user would have to install a library for the plugin to work. We have no mechanisms for this, and will not have. Plugin dependencies beyond what core OpenCPN provides must for these reasons be provided in full by the plugin, in this case by something like libs/xercesc. 

Again the options are to provide a pre-compiled library or to build it from sources.

To build it from source is usually not that complicated. The big obstacle could be dependencies for this build. However, at a glance these looks really simple. Still, it requires cmake code to download sources and build them as required. Example code in _opencpn-libs/marnav/CMakeLists.txt_

Bundling pre-built libraries is possible but so much pain when it comes to maintaining all those linux and macos targets that I would advice against it until the "build from source" path first has been fully explored.

**Comment by Rasbats** [comment]
Now looking at _opencpn-libs/marnav/CMakeLists.txt_

**Comment by leamas** [comment]
There is also the companion _opencpn-libs/flatpak/marnav.yaml_. This corresponds to the snippet you need to add to the flatpak yaml manifest to build the library (as you can see in CMakeLists.txt it basically assumes the that library just exists in the flatpak case, this is because it's listed under _modules:_   in the manifest)

**Comment by Rasbats** [comment]
Still an issue with the includes for Linux builds.
Using this:
 ```
      target_include_directories(XERCES INTERFACE ${xercesc_src_dir})
      message (STATUS "include_directories (XercesC_INCLUDE_DIR: ${xercesc_src_dir})")
```
But get the error:
[https://app.circleci.com/pipelines/github/Rasbats/RTZreader_pi/42/workflows/3eac47b4-c18d-4161-a1e7-4d7e401dbc76/jobs/627](https://app.circleci.com/pipelines/github/Rasbats/RTZreader_pi/42/workflows/3eac47b4-c18d-4161-a1e7-4d7e401dbc76/jobs/627)
Need a clue!

**Comment by Rasbats** [comment]
This is the complete section:
```
# Download sources and get the source directory
      #
      include(FetchContent)
      FetchContent_Declare(xercesc GIT_REPOSITORY ${xercesc_repo} GIT_TAG ${xercesc_tag})
      FetchContent_Populate(xercesc)
      FetchContent_GetProperties(xercesc SOURCE_DIR xercesc_src_dir)

      # Set up build options and build.
      #
      set(BUILD_SHARED_LIBS FALSE CACHE BOOL "Build shared libs" FORCE)
      add_subdirectory(${xercesc_src_dir} xercesc_bld)

      # Add the include files
      #
      target_include_directories(XERCES INTERFACE ${xercesc_src_dir})
      message (STATUS "include_directories (XercesC_INCLUDE_DIR: ${xercesc_src_dir})")

      # Link available libraries
      #
      target_link_libraries(XERCES INTERFACE ocpn::xercesc)
      if (TARGET xercesc::xercesc-io)
        target_link_libraries(XERCES INTERFACE opencpn::xercesc-io)
      endif ()
```
Forget the -io bits.

**Comment by leamas** [comment]
you need to look into that tarball (download and expand it manually) to find out where the headers are.  Perhaps in a include/ sibdirectory or so?

**Comment by leamas** [comment]
... and note that the navmar example had a CMakeLists in the library. Not all have in which case you need to handle for example using autotools instead.  Cmake has provisions to build also such libraries. In short, download the tarball and build it yourself first to get the hang of it.
